pkg/platform/api/buildplanner/model/buildexpression.go

```diff
@@ -97,122 +167,254 @@ type BuildExpression struct {
 	}
 
-	requirements, err := getRequirements(solveNode)
+	ctx := context{}
+	let, err := newLet(ctx, letMap)
 	if err != nil {
-		return nil, errs.Wrap(err, "Could not get requirements node")
+		return nil, errs.Wrap(err, "Could not parse 'let' key")
 	}
 
-	return &BuildExpression{
-		expression:   expression,
-		solveNode:    &solveNode,
-		requirements: requirements,
-	}, nil
-}
-
-// validateRequirements ensures that the requirements in the BuildExpression contain
-// both the name and namespace fields. These fileds are used for requirement operations.
-func validateRequirements(requirements []interface{}) error {
-	for _, requirement := range requirements {
-		r, ok := requirement.(map[string]interface{})
-		if !ok {
-			return errs.New("Requirement in BuildExpression is malformed")
-		}
+	expr := &BuildExpression{let}
 
-		_, ok = r[RequirementNameKey]
-		if !ok {
-			return errs.New("Requirement in BuildExpression is missing name field: %#v", r)
-		}
-		_, ok = r[RequirementNamespaceKey]
-		if !ok {
-			return errs.New("Requirement in BuildExpression is missing namespace field: %#v", r)
-		}
+	err = expr.validateRequirements()
+	if err != nil {
+		return nil, errs.Wrap(err, "Could not validate requirements")
 	}
-	return nil
-}
 
-// Requirements returns the requirements in the BuildExpression.
-func (bx BuildExpression) Requirements() []Requirement {
-	return bx.requirements
+	return expr, nil
 }
 
-// Update updates the BuildExpression's requirements based on the operation and requirement.
-func (bx *BuildExpression) Update(operation Operation, requirement Requirement, timestamp strfmt.DateTime) error {
-	var err error
-	switch operation {
-	case OperationAdded:
-		err = bx.addRequirement(requirement)
-	case OperationRemoved:
-		err = bx.removeRequirement(requirement)
-	case OperationUpdated:
-		err = bx.updateRequirement(requirement)
-	default:
-		return errs.New("Unsupported operation")
+func newLet(ctx context, m map[string]interface{}) (*Let, error) {
+	ctx.push(ctxLet)
+	defer ctx.pop()
+
+	inValue, ok := m["in"]
+	if !ok {
+		return nil, errs.New("Build expression's 'let' object has no 'in' key")
 	}
+
+	in, err := newIn(ctx, inValue)
 	if err != nil {
-		return errs.Wrap(err, "Could not update BuildExpression's requirements")
+		return nil, errs.Wrap(err, "Could not parse 'in' key's value: %v", inValue)
 	}
 
-	formatted, err := time.Parse(time.RFC3339, timestamp.String())
+	// Delete in so it doesn't get parsed as an assignment.
+	delete(m, "in")
+
+	assignments, err := newAssignments(ctx, m)
 	if err != nil {
-		return errs.Wrap(err, "Could not parse latest timestamp")
+		return nil, errs.Wrap(err, "Could not parse 'let' key")
 	}
 
-	(*bx.solveNode)[AtTimeKey] = formatted
+	return &Let{Assignments: *assignments, In: in}, nil
+}
+
+func isAp(ctx context, value map[string]interface{}) bool {
+	ctx.push(ctxIsAp)
+	defer ctx.pop()
 
-	return nil
+	_, hasIn := value["in"]
+	if hasIn && !ctx.contains(ctxAssignments) {
```

Comment:
Is this check dependent on order? i.e. is there a certain key processing order that would cause an invalid result? I don't know enough about build expressions to determine this, but the fact that you're dependent on the stack for the check leads me to think that pushing/popping at unexpected times may lead to unexpected results.

Reply:
I _believe_ that build expressions only support function calls as a result of an assignment, so something like `runtime = solve()` is what we're checking for. You're right that it would fail if there were unexpected pushes and pops but I don't think that's possible. I will double-check that my assumption is correct though.